### PR TITLE
Unqualified host:port pairs are valid Docker auth fields

### LINKED ISSUE #20667: Docker pull secrets with a registry entry of form "host:port" don't get loaded on the node correctly
`credentialprovider.BasicDockerKeyring` tries to do a URL parse on the map key, but the value of the map can be `HOST:PORT`, which results in the returned url having a scheme of `HOST` and an opaque section of `PORT`, instead of that `Host` and `Path` like the method accepts.

This blocks using Docker pull secrets of that form from being added to the map. Workaround is that someone has to set the docker pull secret to contain host entries of `https://HOST:PORT` instead.

### PR DESCRIPTION
**Pull request by smarterclayton**
The dockercfg and .docker/config.json files can contain host:path
combos, which are not correctly parsed by the keyring.

Fixes #20667

### CODE REVIEW

**Comment by smarterclayton** [comment]
Pushed an update - the fix corrected one behavior (parsing was wrong) but didn't remove the previous "workaround" for parsing being wrong, so it created two entries in the list to search instead of one.  Fixed, corrected one test.

**Comment by smarterclayton** [comment]
Had to update the PR - legacy behavior of the hub was not being handled correctly (which has been corrected and tests have been added). PTAL

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #21491

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #21463

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #21463

**Comment by smarterclayton** [comment]
@erictune want your review again since this changed more

**Comment by smarterclayton** [comment]
@deads2k can you review this as well

File: pkg/credentialprovider/keyring.go [hunk 1]
```diff
@@ -77,17 +81,20 @@ func (dk *BasicDockerKeyring) Add(cfg DockerConfig) {
...
-			key := parsed.Host + parsed.Path
-			dk.creds[key] = append(dk.creds[key], creds)
-			dk.index = append(dk.index, key)
+		var key string
+		if (len(effectivePath) > 0) && (effectivePath != "/") {
+			key = parsed.Host + effectivePath
+		} else {
+			key = parsed.Host
 		}
+		dk.creds[key] = append(dk.creds[key], creds)
```
Comment (deads2k) [inline pkg/credentialprovider/keyring.go]: Previously if I had credentials for `host.name/path` would get a map that included 
```
host.name: cred
host.name/path: cred
```

If I'm reading correctly, I now get
```
host.name/path: cred
```


I think that's an intentional change to tighten credential usage.  Please confirm

**Comment by deads2k** [comment]
one question, otherwise lgtm.

File: pkg/credentialprovider/keyring.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/credentialprovider/keyring.go]: Previously, /v1/ and /v2/ were being treated as answering to `host.name`, `host.name/v1/` and `host.name/v2/`.  That means that an auth url like "host.name/v2/some/repo" would be resolved to matching `host.name`, which is very incorrect (we should never present credentials for that).  This change specifically handles those paths and tightens.

**Comment by deads2k** [comment]
Question answered.

LGTM

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue #21467

**Comment by erictune** [comment]
lgtm.

**Pull request merged**